### PR TITLE
Modernise JavaScript API and align with ecosystem

### PR DESCRIPTION
**Pull request by kasperisager**
Notable changes:

- Use promises where it makes sense.
- Consistently refer to IP addresses as `host`.
- Remove `offset` and `length` from `socket.send()`; just use `buffer.subarray()`.
- Add `stream.localPort` and `stream.localHost` as getters for `stream.socket.address()`.
- Add `socket.trySend()` and `stream.trySend()`.

### CODE REVIEW

**Comment by kasperisager** [comment]
@mafintosh This is ready for review.

File: lib/socket.js [hunk 1]
```diff
@@ -83,52 +94,43 @@ module.exports = class UDXSocket extends events.EventEmitter {
...
     } catch (err) {
       this.emit('error', err)
       return
     }
 
-    this.bound = true
-
-    // nextTicked here for nodejs compat
-    queueMicrotask(() => { if (this.bound) this.emit('listening') })
+    queueMicrotask(() => {
```
Comment (mafintosh) [inline lib/socket.js:118]: We can kill the microtask since we don’t need compat now :)

File: lib/socket.js [hunk 2]
```diff
@@ -83,52 +94,43 @@ module.exports = class UDXSocket extends events.EventEmitter {
...
 
     if (!this._inited) this._init()
 
     try {
-      this._port = binding.udx_napi_socket_bind(this._handle, port, ip)
-      this._ip = ip
+      this._port = binding.udx_napi_socket_bind(this._handle, port, host)
+      this._host = host
     } catch (err) {
       this.emit('error', err)
```
Comment (mafintosh) [inline lib/socket.js:114]: we should just make it throw i think instead, clearer.

File: lib/socket.js [hunk 3]
```diff
@@ -156,45 +158,42 @@ module.exports = class UDXSocket extends events.EventEmitter {
     return binding.udx_napi_socket_send_buffer_size(this._handle, size)
   }
 
-  send (buffer, offset, length, port, ip, ttl, onflush) {
-    if (typeof ttl === 'function') {
-      onflush = ttl
-      ttl = 0
-    }
+  async send (buffer, port, host, ttl) {
+    // TODO: should this be an error?
```
Comment (mafintosh) [inline lib/socket.js:162]: i think it makes sense, it didn't send. trySend is for when you don't care

File: lib/socket.js [hunk 4]
```diff
@@ -156,45 +158,42 @@ module.exports = class UDXSocket extends events.EventEmitter {
...
-    return this._reqs.push({ handle, buffer: null, onflush: null }) - 1
+    binding.udx_napi_socket_send_ttl(this._handle, req.handle, id, buffer, port, host, ttl || 0)
+
+    return promise
   }
 
-  static createSocket () {
-    return new this()
+  trySend (buffer, port, host, ttl) {
+    this.send(buffer, port, host, ttl).catch(noop)
```
Comment (mafintosh) [inline lib/socket.js:185]: at some point, it would be nice if trySend could avoid allocating a promise all together to help with gc pressure

**Pull request merged**